File: video_assembler.py
#!/usr/bin/env python3
"""
Naija Creator Hub - Production Video Assembly & Editing Pipeline
================================================================
This script matches stock video clips to a voiceover script, synchronizes
the voiceover timeline, overlays CapCut-style animated word-by-word active 
highlight captions, and compiles them into a single high-fidelity MP4.

Requirements:
    pip install moviepy pillow numpy

Usage:
    python video_assembler.py --script "Script content..." --audio voiceover.wav --videos clip1.mp4 clip2.mp4 --orientation vertical --output output.mp4
"""

import os
import sys
import argparse
from PIL import Image, ImageDraw, ImageFont
import numpy as np

try:
    from moviepy.editor import VideoFileClip, AudioFileClip, concatenate_videoclips, CompositeVideoClip
except ImportError:
    print("[!] Python module 'moviepy' is not installed. Run: pip install moviepy")
    sys.exit(1)

def parse_args():
    parser = argparse.ArgumentParser(description="Naija Creator Hub Video Assembler")
    parser.add_argument("--script", type=str, required=True, help="Full text of the generated script")
    parser.add_argument("--audio", type=str, required=True, help="Path to the generated voiceover audio file (WAV/MP3)")
    parser.add_argument("--videos", nargs="+", required=True, help="List of file paths to the sourced stock video clips")
    parser.add_argument("--orientation", type=str, choices=["vertical", "horizontal"], default="vertical", help="Output format orientation")
    parser.add_argument("--output", type=str, default="final_output.mp4", help="Path to write the final compiled MP4 file")
    parser.add_argument("--font_size", type=int, default=50, help="Base size font for rendering captions")
    parser.add_argument("--highlight_color", type=str, default="#facc15", help="Hex color code for highlighted words (default: yellow)")
    return parser.parse_args()

def split_script_to_sentences(script_text):
    """Splits full script text into sentence segments."""
    import re
    # Split text by punctuation marks with clean spacing
    raw_sentences = re.split(r"(?<=[.!?])\s+|\n+", script_text)
    return [s.strip() for s in raw_sentences if len(s.strip()) > 3]

def generate_word_timestamps(sentences, total_audio_duration):
    """
    Synthesizes and calculates word-level timestamps character-proportionally
    over the total audio voiceover duration.
    """
    total_chars = sum(len(sentence) for sentence in sentences)
    segments = []
    elapsed_time = 0.0

    for i, sentence in enumerate(sentences):
        char_ratio = len(sentence) / max(1, total_chars)
        segment_duration = char_ratio * total_audio_duration
        start_time = elapsed_time
        end_time = elapsed_time + segment_duration
        elapsed_time = end_time

        # Generate word-level timing details
        words = sentence.split()
        words_total_chars = sum(len(w) for w in words)
        word_elapsed = start_time

        timed_words = []
        for word in words:
            word_ratio = len(word) / max(1, words_total_chars)
            word_duration = word_ratio * segment_duration
            word_start = word_elapsed
            word_end = word_elapsed + word_duration
            word_elapsed = word_end

            timed_words.append({
                "text": word,
                "start": word_start,
                "end": word_end
            })

        segments.append({
            "id": i,
            "text": sentence,
            "start": start_time,
            "end": end_time,
            "words": timed_words
        })

    return segments

def draw_caption_on_frame(frame, timestamp, active_segment, width, height, font_size, highlight_color):
    """
    Uses Pillow to dynamically draw CapCut-style dual-toned overlay 
    subtitles onto each raw video frame.
    """
    # Convert moviepy frame (numpy array) to Pillow Image
    img = Image.fromarray(frame)
    draw = ImageDraw.Draw(img)

    # Load font fallback
    try:
        font = ImageFont.truetype("arial.ttf", font_size)
    except IOError:
        try:
            font = ImageFont.truetype("LiberationSans-Bold.ttf", font_size)
        except IOError:
            font = ImageFont.load_default()

    words = active_segment["words"]
    active_word = None
    for w in words:
        if w["start"] <= timestamp <= w["end"]:
            active_word = w
            break

    # CapCut Caption styling arrangement: wrap in rows of 5 words
    max_words_per_row = 5
    words_total = len(words)
    
    # Locate active index
    active_idx = -1
    if active_word:
        try:
            active_idx = words.index(active_word)
        except ValueError:
            pass

    row_index = max(0, active_idx // max_words_per_row) if active_idx != -1 else 0
    row_words = words[row_index * max_words_per_row:(row_index + 1) * max_words_per_row]

    # Compute horizontal spacing
    word_spacing = max_words_per_row * 3
    row_width = 0
    word_widths = []

    for word_obj in row_words:
        # Get width using textbbox
        bbox = draw.textbbox((0, 0), word_obj["text"], font=font)
        w_width = bbox[2] - bbox[0]
        word_widths.append(w_width)
        row_width += w_width

    row_width += word_spacing * (len(row_words) - 1)
    
    # Centered bottom placing
    cap_x = (width / 2) - (row_width / 2)
    cap_y = height * 0.82

    # Draw word strings in series
    current_x = cap_x
    for i, word_obj in enumerate(row_words):
        is_highlighted = (word_obj == active_word)
        color = highlight_color if is_highlighted else "#ffffff"
        txt = word_obj["text"]

        # Draw beautiful high-contrast dark border outlines (standard readable captions)
        border_offsets = [(-2, -2), (2, -2), (-2, 2), (2, 2), (-1, 0), (1, 0), (0, -1), (0, 1)]
        for offset_x, offset_y in border_offsets:
            draw.text((current_x + offset_x, cap_y + offset_y), txt, font=font, fill="#000000")

        # Draw primary fill content
        draw.text((current_x, cap_y), txt, font=font, fill=color)
        current_x += word_widths[i] + word_spacing

    return np.array(img)

def assemble_video():
    args = parse_args()

    print("[+] Naija Creator Hub Video Editor initialized.")
    print(f"[+] Orientation profile: {args.orientation.upper()}")

    # 1. Inspect and load core voiceover sound file
    if not os.path.exists(args.audio):
        print(f"[-] Voiceover audio path not found: {args.audio}")
        sys.exit(1)

    audio_clip = AudioFileClip(args.audio)
    total_duration = audio_clip.duration
    print(f"[+] Voiceover audio file loaded successfully. Play duration: {total_duration:.2f} seconds")

    # 2. Extract scripts sectors
    sentences = split_script_to_sentences(args.script)
    if not sentences:
        print("[-] Script content is empty or lacks valid structure.")
        sys.exit(1)
    print(f"[+] Synced script parsed into {len(sentences)} key narration scenes.")

    # 3. Formulate synchronous timeline mapping words characterized by timescale
    timeline = generate_word_timestamps(sentences, total_duration)

    # Set export bounds matching aspect standards
    width, height = (1080, 1920) if args.orientation == "vertical" else (1920, 1080)

    # 4. Lay and sync video sequences onto the timed segments
    scene_clips = []
    num_available_clips = len(args.videos)

    for i, segment in enumerate(timeline):
        s_start = segment["start"]
        s_end = segment["end"]
        duration = s_end - s_start

        # Fetch index corresponding loops safely
        video_path = args.videos[i % num_available_clips]
        if not os.path.exists(video_path):
            print(f"[-] Video asset file not found at: {video_path}")
            sys.exit(1)

        print(f"    -> Syncing Segment {i+1}/{len(timeline)}: '{segment['text'][:40]}...' to stock asset: {os.path.basename(video_path)}")

        try:
            # Resubmit to moviepy loading container
            clip = VideoFileClip(video_path)

            # Re-scale assets matching export aspect formats using cover cropping
            clip_ratio = clip.w / clip.h
            target_ratio = width / height

            if clip_ratio > target_ratio:
                # Clip is wider than target ratio: crop sides
                new_width = int(clip.h * target_ratio)
                cropped_clip = clip.crop(x1=(clip.w - new_width) // 2, width=new_width)
            else:
                # Clip is taller than target ratio: crop top/bottom
                new_height = int(clip.w / target_ratio)
                cropped_clip = clip.crop(y1=(clip.h - new_height) // 2, height=new_height)

            resized_clip = cropped_clip.resize((width, height))

            # Stretch, loop, or speed up clip if it is shorter than target duration segment
            if resized_clip.duration < duration:
                # Direct looping mapping segment duration limits
                loop_factor = int(np.ceil(duration / resized_clip.duration))
                from moviepy.video.fx.all import loop
                timed_clip = loop(resized_clip, n=loop_factor).subclip(0, duration)
            else:
                timed_clip = resized_clip.subclip(0, duration)

            # Apply dynamic caption overlays to video using Frame manipulation
            def caption_overlay_filter(get_frame, t):
                frame_at_t = get_frame(t)
                # Pass the global timeline coordinate (segment_start + frame_local_t) to get correct active word timing
                global_time = s_start + t
                return draw_caption_on_frame(
                    frame_at_t, 
                    global_time, 
                    segment, 
                    width, 
                    height, 
                    args.font_size, 
                    args.highlight_color
                )

            annotated_clip = timed_clip.fl(caption_overlay_filter)
            scene_clips.append(annotated_clip)

        except Exception as e:
            print(f"[-] Error loading video asset {video_path}: {e}")
            sys.exit(1)

    # 5. Concatenate video clips and bind the master audio timeline
    print("[+] Concatenating and packaging full timeline scenes...")
    assembled_film = concatenate_videoclips(scene_clips, method="compose")
    
    # Bundle voiceover audio overlay track
    final_film = assembled_film.set_audio(audio_clip)

    # 6. Save final file output
    print(f"[+] Rendering and exporting file onto {args.output}...")
    final_film.write_videofile(
        args.output,
        codec="libx264",
        audio_codec="aac",
        fps=30,
        preset="medium",
        threads=4
    )

    print(f"[+] SUCCESS! Your premium compiled video is ready: {args.output}")

if __name__ == "__main__":
    assemble_video()

File: video_assembler.js
#!/usr/bin/env node
/**
 * Naija Creator Hub - Javascript Video Assembly & FFmpeg pipeline
 * =============================================================
 * Handles video mapping, splits narrator segments, outputs timed .ass captions,
 * and streams compilation into a single MP4 file utilizing FFmpeg commands.
 * 
 * Requirements:
 *    npm install fluent-ffmpeg
 *    Assumes ffmpeg is installed on system paths.
 * 
 * Usage:
 *    node video_assembler.js --script "Script data..." --audio voiceover.wav --videos video1.mp4 video2.mp4 --output final.mp4 --orientation vertical
 */

import fs from 'fs';
import path from 'path';
import { exec } from 'child_process';
import https from 'https';

// Mood of the script mapping lists
const extractMoodFromScript = (text) => {
  const content = text.toLowerCase();
  if (content.includes('calm') || content.includes('peace') || content.includes('relax') || content.includes('nature') || content.includes('breathe') || content.includes('soothing')) {
    return 'calm';
  }
  if (content.includes('upbeat') || content.includes('happy') || content.includes('joy') || content.includes('fun') || content.includes('exciting') || content.includes('bright')) {
    return 'upbeat';
  }
  if (content.includes('dramatic') || content.includes('epic') || content.includes('scary') || content.includes('danger') || content.includes('sad') || content.includes('dark')) {
    return 'dramatic';
  }
  if (content.includes('tech') || content.includes('future') || content.includes('cyber') || content.includes('space') || content.includes('cyberpunk')) {
    return 'tech';
  }
  if (content.includes('corporate') || content.includes('business') || content.includes('professional') || content.includes('office') || content.includes('presentation')) {
    return 'corporate';
  }
  return 'motivational';
};

const BGM_TRACKS = {
  motivational: 'https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3',
  dramatic: 'https://www.soundhelix.com/examples/mp3/SoundHelix-Song-2.mp3',
  calm: 'https://www.soundhelix.com/examples/mp3/SoundHelix-Song-3.mp3',
  upbeat: 'https://www.soundhelix.com/examples/mp3/SoundHelix-Song-4.mp3',
  corporate: 'https://www.soundhelix.com/examples/mp3/SoundHelix-Song-5.mp3',
  tech: 'https://www.soundhelix.com/examples/mp3/SoundHelix-Song-8.mp3'
};

const downloadFile = (url, dest) => {
  return new Promise((resolve, reject) => {
    const file = fs.createWriteStream(dest);
    https.get(url, (response) => {
      if (response.statusCode >= 300 && response.statusCode < 400 && response.headers.location) {
        // Handle redirect
        https.get(response.headers.location, (redirResponse) => {
          redirResponse.pipe(file);
          file.on('finish', () => {
            file.close();
            resolve(dest);
          });
        }).on('error', (err) => {
          try { fs.unlinkSync(dest); } catch(e){}
          reject(err);
        });
        return;
      }
      response.pipe(file);
      file.on('finish', () => {
        file.close();
        resolve(dest);
      });
    }).on('error', (err) => {
      try { fs.unlinkSync(dest); } catch(e){}
      reject(err);
    });
  });
};

// Simple arguments parser
const args = {};
process.argv.slice(2).forEach((arg, index, arr) => {
  if (arg.startsWith('--')) {
    const key = arg.slice(2);
    const value = arr[index + 1];
    if (value && !value.startsWith('--')) {
      if (key === 'videos') {
        args[key] = [];
        let i = index + 1;
        while (arr[i] && !arr[i].startsWith('--')) {
          args[key].push(arr[i]);
          i++;
        }
      } else {
        args[key] = value;
      }
    }
  }
});

const script = args.script || '';
const audioPath = args.audio || '';
const videoPaths = args.videos || [];
const orientation = args.orientation || 'vertical';
const outputPath = args.output || 'compiled_output.mp4';
const highlightColor = args.highlight_color || '00FFFF'; // ASS style: BGR hex (Yellow custom is 00FFFF)

if (!script || !audioPath || videoPaths.length === 0) {
  console.log('[!] Usage: node video_assembler.js --script "your script..." --audio voiceover.wav --videos clip1.mp4 clip2.mp4 --output final.mp4');
  console.log('[!] Ensure to add appropriate arguments.');
  process.exit(1);
}

// 1. Get Audio duration using ffprobe
const getMediaDuration = (filePath) => {
  return new Promise((resolve, reject) => {
    exec(`ffprobe -v error -show_entries format=duration -of default=noprint_wrappers=1:nokey=1 "${filePath}"`, (err, stdout) => {
      if (err) return reject(err);
      resolve(parseFloat(stdout.trim()));
    });
  });
};

const assemble = async () => {
  try {
    console.log('[+] Starting Javascript FFmpeg video editor...');
    const audioDuration = await getMediaDuration(audioPath);
    console.log(`[+] Input audio duration details calculated: ${audioDuration} seconds`);

    // Split sentences
    const sentences = script
      .split(/(?<=[.!?])\s+|\n+/)
      .map(s => s.trim())
      .filter(s => s.length > 3);

    if (sentences.length === 0) {
      console.log('[-] Invalid script content.');
      process.exit(1);
    }

    const totalChars = sentences.reduce((sum, s) => sum + s.length, 0);

    // Build timeline mapping
    let elapsed = 0;
    const timeline = sentences.map((sentence, idx) => {
      const weight = sentence.length / totalChars;
      const duration = weight * audioDuration;
      const start = elapsed;
      const end = elapsed + duration;
      elapsed = end;

      // Group words
      const words = sentence.split(/\s+/);
      const totalWordChars = words.reduce((acc, w) => acc + w.length, 0);
      let wordElapsed = start;
      
      const timedWords = words.map((w) => {
        const wordWeight = w.length / totalWordChars;
        const wDur = wordWeight * duration;
        const wStart = wordElapsed;
        const wEnd = wordElapsed + wDur;
        wordElapsed = wEnd;
        return { text: w, start: wStart, end: wEnd };
      });

      return {
        id: idx,
        text: sentence,
        start,
        end,
        words: timedWords
      };
    });

    // Write premium Advanced SubStation Alpha (.ass) style subtitle cue entries
    // This allows custom styles: fonts, sizes, strokes, margins and word-by-word karaoke highlights!
    const assFile = path.resolve('./temp_subtitles.ass');
    const assHeader = `[Script Info]
Title: Creator Hub Auto Subtitles
ScriptType: v4.00+
PlayResX: 1080
PlayResY: 1920

[V4+ Styles]
Format: Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, OutlineColour, BackColour, Bold, Italic, Underline, StrikeOut, ScaleX, ScaleY, Spacing, Angle, BorderStyle, Outline, Shadow, Alignment, MarginL, MarginR, MarginV, Encoding
Style: CapCut,Space Grotesk,50,&H00FFFFFF,&H000000FF,&H00000000,&H00000000,-1,0,0,0,100,100,0,0,1,5,1,2,50,50,300,1

[Events]
Format: Layer, Start, End, Style, Name, MarginL, MarginR, MarginV, Effect, Text
`;

    let assBody = '';
    const formatAssTime = (sec) => {
      const hrs = Math.floor(sec / 3600);
      const mins = Math.floor((sec % 3600) / 60);
      const secs = Math.floor(sec % 60);
      const ms = Math.floor((sec % 1) * 100);
      return `${hrs}:${mins.toString().padStart(2, '0')}:${secs.toString().padStart(2, '0')}.${ms.toString().padStart(2, '0')}`;
    };

    // Generating timing dialogues word-by-word highlights
    timeline.forEach(segment => {
      segment.words.forEach(wordObj => {
        const startStr = formatAssTime(wordObj.start);
        const endStr = formatAssTime(wordObj.end);
        
        // Wrap other words of sentence as standard white, only active word colored
        const styledSentence = segment.words.map(w => {
          if (w === wordObj) {
            return `{\\1c&H${highlightColor}&}${w.text}{\\1c&HFFFFFF&}`;
          }
          return w.text;
        }).join(' ');

        assBody += `Dialogue: 0,${startStr},${endStr},CapCut,,0,0,0,,${styledSentence}\n`;
      });
    });

    fs.writeFileSync(assFile, assHeader + assBody);
    console.log('[+] Advanced timed sub-captions file written successfully.');

    // Auto-detect script mood theme & download background track loop safely
    const detectedMood = extractMoodFromScript(script);
    const bgmUrl = BGM_TRACKS[detectedMood] || BGM_TRACKS.motivational;
    const bgmTempPath = path.join(path.dirname(outputPath) || '.', 'temp_bgm.mp3');

    console.log(`[+] Auto-detected script mood theme: ${detectedMood.toUpperCase()}`);
    console.log(`[+] Sourcing background music loops from: ${bgmUrl}`);
    
    try {
      await downloadFile(bgmUrl, bgmTempPath);
      console.log(`[+] Background loop track prepared and cached successfully.`);
    } catch (bgmErr) {
      console.error(`[-] Could not download BGM track loop, running voiceover compile fallback:`, bgmErr);
    }

    const hasBgm = fs.existsSync(bgmTempPath);

    // 2. Generate FFmpeg command scripts
    const w = orientation === 'vertical' ? 1080 : 1920;
    const h = orientation === 'vertical' ? 1920 : 1080;

    // Concatenate video filters dynamically in a single pipeline
    // For each segment index, feed the matching sourced clip scaled, cropped to target orientation box
    console.log('[+] Preparing ffmpeg command chains matching timeline mapping window...');
    let filterComplex = '';
    let inputs = '';

    videoPaths.forEach((vPath, idx) => {
      inputs += ` -i "${vPath}"`;
    });
    // Add voiceover audio input
    inputs += ` -i "${audioPath}"`;
    
    if (hasBgm) {
      // Add background music loop input
      inputs += ` -stream_loop -1 -i "${bgmTempPath}"`;
    }

    // Map each segment slot in timeline in order
    let sceneInputsCount = videoPaths.length;
    let bmgInputIdx = sceneInputsCount + 1; // index in FFmpeg inputs array
    let mapStrs = '';

    timeline.forEach((seg, index) => {
      const clipIdx = index % sceneInputsCount;
      const duration = seg.end - seg.start;
      
      // Scale and Crop to vertical or landscape
      filterComplex += `[${clipIdx}:v]scale=w=${w}:h=${h}:force_original_aspect_ratio=increase,crop=${w}:${h},trim=duration=${duration},setpts=PTS-STARTPTS[v${index}];`;
      mapStrs += `[v${index}]`;
    });

    filterComplex += `${mapStrs}concat=n=${timeline.length}:v=1:a=0[vmaster];`;
    
    if (hasBgm) {
      // Attach Subtitles track filter
      filterComplex += `[vmaster]subtitles='${assFile.replace(/\\/g, '/')}'[vfinal];`;
      // Audio Mixer Graph
      filterComplex += `[${sceneInputsCount}:a]volume=1.0[vo];[${bmgInputIdx}:a]volume=0.15[bg];[vo][bg]amix=inputs=2:duration=first:dropout_transition=2[mixed_audio]`;
    } else {
      filterComplex += `[vmaster]subtitles='${assFile.replace(/\\/g, '/')}'[vfinal]`;
    }

    const finalCmd = hasBgm
      ? `ffmpeg -y ${inputs} -filter_complex "${filterComplex}" -map "[vfinal]" -map "[mixed_audio]" -c:v libx264 -preset medium -crf 23 -c:a aac -b:a 192k "${outputPath}"`
      : `ffmpeg -y ${inputs} -filter_complex "${filterComplex}" -map "[vfinal]" -map ${sceneInputsCount}:a -c:v libx264 -preset medium -crf 23 -c:a aac -b:a 192k "${outputPath}"`;

    console.log('[+] Run command pipeline generated:');
    console.log(`    ${finalCmd.slice(0, 500)}...`);

    console.log('[+] Compiling final media clips using FFmpeg wrapper. Please wait...');
    
    exec(finalCmd, (execErr, stdout, stderr) => {
      // Clean temporary timing subtitles and temporary download tracks
      try { fs.unlinkSync(assFile); } catch(e){}
      try { if (hasBgm) fs.unlinkSync(bgmTempPath); } catch(e){}

      if (execErr) {
        console.error('[-] Compilation error: ', execErr);
        console.log('[-] Ensure ffmpeg is installed and subtitle directories/configuration parameters are correct.');
        process.exit(1);
      }
      console.log(`[+] SUCCESS! Final production movie exported successfully: ${outputPath}`);
    });

  } catch (error) {
    console.error('[-] Error executing Javascript assembly pipeline: ', error);
  }
};

assemble();
